File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = defaultdict(int)

        for i in nums:
            count[i] += 1

        bucket = [[] for _ in range(len(nums) + 1)]

        for j in count:
            bucket[count[j]].append(j)

        res = []

        for n in range(len(bucket) - 1, -1, -1):
            for i in bucket[n]:
                res.append(i)
                if len(res) == k:
                    return res

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int n: nums){
            map.put(n, map.getOrDefault(n, 0) + 1);
        }

        List<Integer>[] count = new List[nums.length + 1];
        for (int key: map.keySet()){
            if(count[map.get(key)] == null){
                count[map.get(key)] = new ArrayList<>();
            }
            count[map.get(key)].add(key);
        }
        int[] result = new int[k];
        int counter = 0;
        for(int i = nums.length; i >= 0 && counter < k; i--){
            if (count[i] != null ){
                for (int j : count[i]){
                    result[counter++] = j;
                    if (counter == k) { break; }
                }
            }
        }

        return result;

        }

    }